route_auth.go

```diff
@@ -1,102 +1,242 @@
 package main
 
 import (
+	"bytes"
 	"errors"
 	"fmt"
-	"net/http"
-	"math/rand"
+	"net/url"
+	"time"
 
+	"github.com/go-playground/validator/v10"
 	"github.com/pdeguing/empire-and-foundation/data"
 	"github.com/pdeguing/empire-and-foundation/ent"
-	"github.com/pdeguing/empire-and-foundation/ent/user"
 	"github.com/pdeguing/empire-and-foundation/ent/planet"
+	"github.com/pdeguing/empire-and-foundation/ent/user"
 	"golang.org/x/crypto/bcrypt"
+	"html/template"
+	"math/rand"
+	"net/http"
 )
 
 // GET /signup
 // Show the signup page
 func serveSignup(w http.ResponseWriter, r *http.Request) {
 	generateHTML(w, r, "signup", nil, "layout", "public.navbar", "flash", "signup")
 	forgetForm(r)
+	forgetFormErrors(r)
+}
+
+type signupAccountRequest struct {
+	Email          string `json:"email" name:"email" validate:"required,email"`
+	Username       string `json:"username" name:"username" validate:"required,min=2,max=30,unique_username"`
+	Password       string `json:"password" name:"password" validate:"required,min=8"`
+	PasswordRepeat string `json:"password_confirm" name:"confirm password" validate:"required,eqfield=Password"`
 }
 
-// POST /signup
+// POST /signup_account
 // Create the user account
 func serveSignupAccount(w http.ResponseWriter, r *http.Request) {
-	err := r.ParseForm()
-	if err != nil {
+	forgetFormErrors(r)
+	var input signupAccountRequest
+	if err := r.ParseForm(); err != nil {
 		serveError(w, r, newInternalServerError(fmt.Errorf("unable to parse form: %v", err)))
 		return
 	}
-	password, err := bcrypt.GenerateFromPassword([]byte(r.PostFormValue("password")), 14)
+	if err := decoder.Decode(&input, r.PostForm); err != nil {
```

Comment:
Why do we need to decode the input now?

Reply:
Then I can use the validator on it. The rest of the code is also less error prone because it won't compile if I try to access a field that doesn't exist.